core/templates/dev/head/filters.js

```diff
@@ -268,6 +268,50 @@ oppia.filter('normalizeWhitespace', [function() {
   };
 }]);
 
+// Filter that takes a string, trims and normalizes spaces within each
+// line, and removes blank lines. Note that any spaces whose removal does not
+// result in two alphanumeric "words" being joined together are also removed,
+// so "hello ? " becomes "hello?".
```

Comment:
The filter doesn't remove punctuation, but it removes spaces around punctuation. This is more aggressive than the usual normalizeSpaces filter.

Maybe I should call this normalizeWhitespaceAndCaseAggressively? I'm not sure...